tensorflow_recommenders_addons/dynamic_embedding/python/ops/layers.py

```diff
@@ -0,0 +1,378 @@
+    partitioner = kwargs.get('partitioner', devar.default_partition_fn)
+    trainable = kwargs.get('trainable', True)
+    max_norm = kwargs.get('max_norm', None)
+
+    parameter_name = name + '-parameter' if name else 'EmbeddingParameter'
+    with tf.name_scope('DynamicEmbedding'):
+      self.params = de.get_variable(parameter_name,
+                                    key_dtype=key_dtype,
+                                    value_dtype=value_dtype,
+                                    dim=self.embedding_size,
+                                    devices=devices,
+                                    partitioner=partitioner,
+                                    shared_name='layer_embedding_variable',
+                                    initializer=initializer,
+                                    trainable=trainable,
+                                    checkpoint=True,
+                                    init_size=kwargs.get('init_capacity', 0),
+                                    kv_creator=kwargs.get('kv_creator', None),
+                                    restrict_policy=kwargs.get(
+                                        'restrict_policy', None),
+                                    bp_v2=kwargs.get('bp_v2', True))
+
+      distribute_strategy = kwargs.get('distribute_strategy', None)
+      shadow_name = name + '-shadow' if name else 'ShadowVariable'
+      self.shadow = de.shadow_ops.ShadowVariable(
+          self.params,
+          name=shadow_name,
+          max_norm=max_norm,
+          trainable=trainable,
+          distribute_strategy=distribute_strategy)
+    self._current_ids = self.shadow.ids
+    self._current_exists = self.shadow.exists
+    self.optimizer_vars = self.shadow._optimizer_vars
+    super(Embedding, self).__init__(name=name,
+                                    trainable=trainable,
+                                    dtype=value_dtype)
+
+  @tf.function
+  def call(self, ids):
+    """
+    Make feature ids input becomes a embedding tensor. There are two stages
+    here. The first stage will lookup the dyamic_embedding.Variable and get
+    raw output of the embedding with shape (shape(ids), embedding_size).
+    The second stage will squash the raw embedding output to a fixed shape
+    embedding. Here the fixed shape is (batch_size, embedding_size). We
+    need a fixed shape here because the keras layer requires a explicitly
+    set shape in output.
+
+    It could be override in an inherited class from Embedding, to achieve
+    a custom embedding layer.
+
+    Args:
+      ids: feature ids of the input. It should be same dtype as the key_dtype
+        of the layer.
+
+    Returns:
+      A embedding output with shape (batch_size, embedding_size) if the input
+        are batched, else its shape is (embedding_size).
+    """
+    ids = tf.convert_to_tensor(ids)
+    input_shape = tf.shape(ids)
+    lookup_result = de.shadow_ops.embedding_lookup(self.shadow, ids)
```

Comment:
embedding_lookup_unique is a better choice for following Keras Embedding layer?

Reply:
It does! In dynamic_embedding_ops.py, the `xxx_embedding_lookup` APIs are wrappers of `embedding_lookup`. So I'm trying to reuse the codes in these wrappers, by applying code transform on `de.embedding_lookup`. It will take some time, maybe just making it `unique_embedding_lookup` on current keras layer is realistic.